docs/settings.rst
===================
配置参数
===================

.. contents::
    :local:
    :depth: 1


Django Cool 在 Django 中的配置都在 `DJANGO_COOL` 中。如：在 `settings.py` 中设置：

.. code-block:: python

    DJANGO_COOL = {
        'API_ERROR_CODES': (
            ('ERR_TEST1', (10001, 'test1错误码')),
            ('ERR_TEST2', (10002, 'test2错误码')),
            ('ERR_TEST3', (10003, 'test3错误码')),
        ),
    }


Admin
====================

:mod:`cool.admin` 配置

.. setting:: MODEL_SET_VERBOSE_NAME_TO_DB_COMMENT

``MODEL_SET_VERBOSE_NAME_TO_DB_COMMENT``
---------------------------------------------------------------
默认值： ``False``

设置为 ``True`` 当 model 的 field 未设置 `db_comment` 时, 自动将 `db_comment` 设置为 `verbose_name` 的值

.. setting:: MODEL_SET_VERBOSE_NAME_TO_DB_COMMENT

``MODEL_SET_DEFAULT_TO_DB_DEFAULT``
---------------------------------------------------------------
默认值： ``False``

设置为 ``True`` 当 model 的 field 未设置 `db_default` 时, 自动将 `db_default` 设置为 `default` 的值


Admin
====================

:mod:`cool.admin` 配置

.. setting:: ADMIN_AUTOCOMPLETE_CHECK_PERM

``ADMIN_AUTOCOMPLETE_CHECK_PERM``
---------------------------------------------------------------
默认值： ``True``

自动提示组件是否判断读权限，设置为 ``True`` 用户需要有读权限才能在下拉组件中列出数据，设置为 ``False`` 只需要登陆就可以列出数据

.. setting:: ADMIN_FILTER_USE_SELECT

``ADMIN_FILTER_USE_SELECT``
---------------------------------------------------------------
默认值： ``True``

筛选器列表展示为下拉框

.. setting:: ADMIN_FILTER_WITH_HUMAN_TITLE

``ADMIN_FILTER_WITH_HUMAN_TITLE``
---------------------------------------------------------------
默认值： ``True``

关联表字段场景筛选器显示内容优化

.. setting:: ADMIN_FOREIGNKEY_FIELD_USE_AUTOCOMPLETE

``ADMIN_FOREIGNKEY_FIELD_USE_AUTOCOMPLETE``
---------------------------------------------------------------
默认值： ``True``

外键字段在admin中默认使用自动提示组件

.. setting:: ADMIN_MANYTOMANY_FIELD_USE_AUTOCOMPLETE

``ADMIN_MANYTOMANY_FIELD_USE_AUTOCOMPLETE``
---------------------------------------------------------------
默认值： ``True``

多对多字段在admin中默认使用自动提示组件展示

.. setting:: ADMIN_RELATED_FIELD_FILTER_USE_AUTOCOMPLETE

``ADMIN_RELATED_FIELD_FILTER_USE_AUTOCOMPLETE``
---------------------------------------------------------------
默认值： ``True``

外键字段在筛选器中默认使用自动提示组件展示

.. setting:: ADMIN_SHOW_IMAGE_IN_CHANGE_LIST

``ADMIN_SHOW_IMAGE_IN_CHANGE_LIST``
---------------------------------------------------------------
默认值： ``True``

图片字段在admin列表中显示图片

.. setting:: ADMIN_SHOW_IMAGE_IN_FORM_PAGE

``ADMIN_SHOW_IMAGE_IN_FORM_PAGE``
---------------------------------------------------------------
默认值： ``True``

图片字段在admin编辑页面中显示图片

.. setting:: ADMIN_DATE_FIELD_FILTER_USE_RANGE

``ADMIN_DATE_FIELD_FILTER_USE_RANGE``
---------------------------------------------------------------
默认值： ``True``

日期字段在筛选器中默认使用日期范围组件展示

.. setting:: ADMIN_SITE_TITLE

``ADMIN_SITE_TITLE``
---------------------------------------------------------------
默认值： ``None``

在每个管理页面的 <title> （字符串）末尾放置的文字

.. setting:: ADMIN_SITE_HEADER

``ADMIN_SITE_HEADER``
---------------------------------------------------------------
默认值： ``None``

要放在每个管理页面顶部的文字，作为 <h1> （一个字符串）

.. setting:: ADMIN_INDEX_TITLE

``ADMIN_INDEX_TITLE``
---------------------------------------------------------------
默认值： ``None``

放在管理索引页顶部的文字（一个字符串）

``ADMIN_THEME``
---------------------------------------------------------------
默认值： ``None``

后台风格模板, 内置模板支持： `DEFAULT`: Django默认 `DARK`: Django黑暗模式 `BLANK`: 黑色风格

``ADMIN_SITE_REGISTER_FILTER_FUNCTION``
---------------------------------------------------------------
默认值： ``None``

admin 注册 model 通用参数处理函数

.. code-block:: python

    def site_register_filter(model_class, *, admin_class, site, **options):
        # 通用 options 处理
        return options

APIView
====================
:mod:`cool.views` 相关配置

.. setting:: API_EXCEPTION_DEFAULT_STATUS_CODE

``API_EXCEPTION_DEFAULT_STATUS_CODE``
---------------------------------------------------------------
默认值： ``400``

``CoolAPIException`` 默认 ``status_code``

.. setting:: API_SYSTEM_ERROR_STATUS_CODE

``API_SYSTEM_ERROR_STATUS_CODE``
---------------------------------------------------------------
默认值： ``500``

未捕获异常 默认 ``status_code``

.. setting:: API_PARAM_ERROR_STATUS_CODE

``API_PARAM_ERROR_STATUS_CODE``
---------------------------------------------------------------
默认值： ``400``

参数验证错误默认 ``status_code``

.. setting:: API_SUCCESS_WITH_CODE_MSG

``API_SUCCESS_WITH_CODE_MSG``
---------------------------------------------------------------
默认值： ``True``

返回成功时，返回结果是否带 ``code`` ``message`` ``data`` 一层, 设置为``False``成功是时只返回``data``中内容

.. setting:: API_SHOW_PARAM_ERROR_INFO

``API_SHOW_PARAM_ERROR_INFO``
---------------------------------------------------------------
默认值： ``True``

参数验证错误时是否返回错误描述

.. setting:: API_SHOW_PARAM_ERROR_INFO

``API_SUCCESS_CODES``
---------------------------------------------------------------
默认值： ``True``

API返回成功时的返回码值，对应 ``ErrorCode.SUCCESS``

.. setting:: API_ERROR_CODES

``API_ERROR_CODES``
---------------------------------------------------------------
默认值： ``()``

自定义错误码列表, 如settings中设置如下后

.. code-block:: python

    DJANGO_COOL = {
        'API_ERROR_CODES': (
            ('ERR_TEST1', (10001, 'test1错误码')),
            ('ERR_TEST2', (10002, 'test2错误码')),
            ('ERR_TEST3', (10003, 'test3错误码')),
        ),
    }


可以使用如下代码使用

.. code-block:: python

    from cool.views import ErrorCode
    print(ErrorCode.ERR_TEST1)
    print(ErrorCode.ERR_TEST2.code)
    print(ErrorCode.ERR_TEST3.desc)


.. setting:: API_DEFAULT_CODE_KEY

``API_DEFAULT_CODE_KEY``
---------------------------------------------------------------
默认值： ``'code'``

API返回内容中返回码键名称

.. setting:: API_DEFAULT_MESSAGE_KEY

``API_DEFAULT_MESSAGE_KEY``
---------------------------------------------------------------
默认值： ``'message'``

API返回内容中描述键名称

.. setting:: API_DEFAULT_DATA_KEY

``API_DEFAULT_DATA_KEY``
---------------------------------------------------------------
默认值： ``'data'``

API返回内容中数据键名称

.. setting:: API_RESPONSE_DICT_FUNCTION

``API_RESPONSE_DICT_FUNCTION``
---------------------------------------------------------------
默认值： ``'cool.views.response.get_response_dict'``

组装返回结果函数

.. code-block:: python

    def get_response_dict(code, message, data, success_with_code_msg, status_code, response_data, **kwargs):
        if not success_with_code_msg and code == ErrorCode.SUCCESS:
            return data
        else:
            return {
                cool_settings.API_DEFAULT_CODE_KEY: code,
                cool_settings.API_DEFAULT_MESSAGE_KEY: message,
                cool_settings.API_DEFAULT_DATA_KEY: data,
            }

.. setting:: API_WS_REQ_ID_NAME

``API_WS_REQ_ID_NAME``
---------------------------------------------------------------
默认值： ``'req_id'``

``CoolBFFAPIConsumer`` 中请求id键名称

.. setting:: API_WS_REQ_PATH_NAME

``API_WS_REQ_PATH_NAME``
---------------------------------------------------------------
默认值： ``'path'``

``CoolBFFAPIConsumer`` 中请求路径键名称

.. setting:: API_WS_REQ_DATA_NAME

``API_WS_REQ_DATA_NAME``
---------------------------------------------------------------
默认值： ``'data'``

``CoolBFFAPIConsumer`` 中请求数据键名称

.. setting:: API_WS_RES_STATUS_CODE_NAME

``API_WS_RES_STATUS_CODE_NAME``
---------------------------------------------------------------
默认值： ``'status_code'``

``CoolBFFAPIConsumer`` 中返回状态码键名称

.. setting:: API_WS_RES_SERVER_TIME_NAME

``API_WS_RES_SERVER_TIME_NAME``
---------------------------------------------------------------
默认值： ``'server_time'``

``CoolBFFAPIConsumer`` 中返回服务器时间键名称

.. setting:: API_WS_RES_DATA_NAME

``API_WS_RES_DATA_NAME``
---------------------------------------------------------------
默认值： ``'data'``

``CoolBFFAPIConsumer`` 中返回数据键名称

.. setting:: API_WS_RES_STATUS_CODE_NOT_FOUND

``API_WS_RES_STATUS_CODE_NOT_FOUND``
---------------------------------------------------------------
默认值： ``404``

``CoolBFFAPIConsumer`` 中找不到接口时返回状态码

.. setting:: API_WS_RES_STATUS_CODE_SERVER_ERROR

``API_WS_RES_STATUS_CODE_SERVER_ERROR``
---------------------------------------------------------------
默认值： ``500``

``CoolBFFAPIConsumer`` 中未捕获异常时返回状态码


docs/changelog.rst
Changelog
================

Version 1.3.4
------------------
+ 调整超链接颜色

Version 1.3.3
------------------
+ Fixed `InfoMixin` 兼容代码错误

Version 1.3.2
------------------
+ `BaseModelAdmin` 中 `heads` `tails` 的配置按顺序展示

Version 1.3.1
------------------
+ Fixed 外键包含空时处理逻辑优化

Version 1.3.0
------------------
+ 增加 `ADMIN_FILTER_WITH_HUMAN_TITLE` 配置， 关联表字段场景筛选器显示内容优化
+ `BaseModelAdmin` 中 `list_display` 出现关联表字段时，标题、内容显示优化
+ 单元测试增加 `Python 3.13` `Django 5.2`
+ `CRIDMixin` 增加 `get_queryset`
+ 抽离 `GetOneMixin` 获取单独信息，供 `InfoMixin` 和 `EditMixin`
+ 抽离 `BaseInfoMixin` 和 `BaseEditMixin` 供不需要传唯一键场景使用（如从登录信息中获取）

Version 1.2.14
------------------
+ `ListSerializer` 支持 `exclude` 参数

Version 1.2.13
------------------
+ `Constants` 支持扩展属性

Version 1.2.12
------------------
+ 增加 `MODEL_SET_DEFAULT_TO_DB_DEFAULT`配置， 当 model 的 field 未设置 `db_default` 时, 自动将 `db_default` 设置为 `default` 的值

Version 1.2.11
------------------
+ 增加 `MODEL_SET_VERBOSE_NAME_TO_DB_COMMENT`配置， 当 model 的 field 未设置 `db_comment` 时, 自动将 `db_comment` 设置为 `verbose_name` 的值
+ Fixed python3.11中 `Constants` 排重逻辑不正确

Version 1.2.10
------------------
+ Django 4.0 删除 `request.get_raw_uri()`, 使用 `request.build_absolute_uri()` 替代

Version 1.2.9
------------------
+ 修复 `verbose_name` 或 `help_text` 为 `lazy_str` 时候 `view`相关接口可能报错

Version 1.2.8
------------------
+ `SearchListMixin` 中 增加 `order_fields` 排序方式
+ 增加 `ListSerializer` 支持 `order_by`, `filter`, `limit` 参数, `BaseSerializer(many=True)`默认支持
+ `CoolBFFAPIView` 增加缓存功能，统计类接口可设置同请求参数接口数据缓存

Version 1.2.4
------------------
+ Fixed `InfoMixin` 中 `pk_id` 为 `False` 时报错
+ 接口文档中支持不展示基础字段

Version 1.2.3
------------------
+ `SearchListMixin` 增加 `filter_fields`

Version 1.2.2
------------------
+ `ADMIN_SITE_REGISTER_FILTER_FUNCTION` 支持 `site` `admin_class` 自定义

Version 1.2.1
------------------
+ 增加 `ADMIN_SITE_REGISTER_FILTER_FUNCTION` admin 注册 model 通用参数处理函数

Version 1.2.0
------------------
+ `CoolBFFAPIView` 增加 `support_methods` 设置请求类型
+ 增加admin后台风格自定义
+ `View` 文档支持 `django.views.generic.base.View`
+ `_MODEL_WITH_CACHE` 设置不使用缓存时减少部分逻辑

Version 1.1.4
------------------
+ `BaseModelAdmin`增加 `auto_set_list_select_related` 参数， 当`list_select_related`为`False`时，自动将`list_display`中外键字段自动写入`list_select_related`

Version 1.1.3
------------------
+ `InfoMixin` 的 `ex_unique_ids`、 `AddMixin` 的 `add_fields`、 `EditMixin` 的 `edit_fields` 字段支持['name', ('school_code', 'school_id)] 的方式设置请求参数和字段名称不一致的情况

Version 1.1.2
------------------
+ Fixed `get_rest_field_from_model_field` 生成外键字段类型错误

Version 1.1.1
------------------
+ `ModelCache` 支持非简单字段

Version 1.1.0
------------------
+ 拆分 `AutoCompleteMixin` 可供 `TabularInline`、`StackedInline`使用
+ `BaseModelAdmin` 是否可修改字段从 `changeable` 修改为 `editable` （2.0 将删除`changeable`字段）

Version 1.0.16
------------------
+ `CoolAutocompleteMixin` 兼容 django3.2
+ `SplitCharField` json 方式提交时，支持list

Version 1.0.15
------------------
+ `ModelCacheMixin` 缓存重构， 支持联合唯一索引数据缓存获取，联合唯一键使用`(key1, key2) in ((val1, val2), (val3, val4))`的方式
+ 测试用例增加不同数据库测试

Version 1.0.13
------------------
+ 增加 `API_SUCCESS_CODE` 配置，定制成功返回时的code
+ `API_RESPONSE_DICT_FUNCTION` 函数增加参数

Version 1.0.12
------------------
+ Fixed mixins 编辑BUG

Version 1.0.11
------------------
+ Fixed 修改扩展字段传空数组时，未做删除处理

Version 1.0.10
------------------
+ 增加 列表、增删改基础API

Version 1.0.9
------------------
+ `get_rest_field_from_model_field` 支持外键字段
+ 子项 extend_info_format 增加 help_text

Version 1.0.8
------------------
+ log_exception 修改为 error 级别
+ log_response 中返回数据顺序修改移后

Version 1.0.7
------------------
+ cache key 为 `None` 时， 生成完成key为 prefix

Version 1.0.6
------------------
+ `ModelFieldChangeMixin` 增加批量操作
+ `RecursiveField` 字段说明父级出现过两次以上就不显示字段说明内容

Version 1.0.5
------------------
+ 增加 `RecursiveField` 支持树树型结构序列化

Version 1.0.4
------------------
+ 增加序列化请求字段 `SerializerField`

Version 1.0.3
------------------
+ `WidgetFilterMixin` 组件防止数据库查询
+ `AbstractUserMixin` 中 `get_username` 优先从父类获取

Version 1.0.2
------------------
+ `get_rest_field_from_model_field` 中 `BooleanField` 默认为 `None` 时候 类型设置为 `NullBooleanField`

Version 1.0.1
------------------
+ Fixed 增加 data 类型判断

Version 1.0.0
------------------
+ 初始化版本


cool/settings.py
# encoding: utf-8

from django.conf import settings
from django.test.signals import setting_changed
from django.utils.module_loading import import_string

DEFAULTS = {
    # Model
    'MODEL_SET_VERBOSE_NAME_TO_DB_COMMENT': False,
    'MODEL_SET_DEFAULT_TO_DB_DEFAULT': False,
    # Admin
    'ADMIN_AUTOCOMPLETE_CHECK_PERM': True,
    'ADMIN_FILTER_USE_SELECT': True,
    'ADMIN_FILTER_WITH_HUMAN_TITLE': True,
    'ADMIN_FOREIGNKEY_FIELD_USE_AUTOCOMPLETE': True,
    'ADMIN_MANYTOMANY_FIELD_USE_AUTOCOMPLETE': True,
    'ADMIN_RELATED_FIELD_FILTER_USE_AUTOCOMPLETE': True,
    'ADMIN_SHOW_IMAGE_IN_CHANGE_LIST': True,
    'ADMIN_SHOW_IMAGE_IN_FORM_PAGE': True,
    'ADMIN_DATE_FIELD_FILTER_USE_RANGE': True,
    'ADMIN_SITE_TITLE': None,
    'ADMIN_SITE_HEADER': None,
    'ADMIN_INDEX_TITLE': None,
    'ADMIN_THEME': None,
    'ADMIN_SITE_REGISTER_FILTER_FUNCTION': None,

    # APIView
    'API_EXCEPTION_DEFAULT_STATUS_CODE': 400,

    'API_SYSTEM_ERROR_STATUS_CODE': 500,
    'API_PARAM_ERROR_STATUS_CODE': 400,
    'API_SUCCESS_WITH_CODE_MSG': True,
    'API_SHOW_PARAM_ERROR_INFO': True,

    'API_SUCCESS_CODE': 0,
    'API_ERROR_CODES': (),

    'API_DEFAULT_CODE_KEY': 'code',
    'API_DEFAULT_MESSAGE_KEY': 'message',
    'API_DEFAULT_DATA_KEY': 'data',

    'API_RESPONSE_DICT_FUNCTION': 'cool.views.response.get_response_dict',

    # websocket
    'API_WS_REQ_ID_NAME': 'req_id',
    'API_WS_REQ_PATH_NAME': 'path',
    'API_WS_REQ_DATA_NAME': 'data',
    'API_WS_RES_STATUS_CODE_NAME': 'status_code',
    'API_WS_RES_SERVER_TIME_NAME': 'server_time',
    'API_WS_RES_DATA_NAME': 'data',
    'API_WS_RES_STATUS_CODE_NOT_FOUND': 404,
    'API_WS_RES_STATUS_CODE_SERVER_ERROR': 500,

}


# List of settings that may be in string import notation.
IMPORT_STRINGS = [
    'API_RESPONSE_DICT_FUNCTION',
    'ADMIN_SITE_REGISTER_FILTER_FUNCTION'
]


def get_admin_theme(val, style=False):
    from cool.admin import theme
    if val is None:
        return None
    if isinstance(val, theme.Theme):
        assert not style
        return val
    elif isinstance(val, theme.ThemeStyle):
        return val if style else theme.Theme(val)
    elif isinstance(val, (list, tuple)):
        assert not style
        if len(val) > 2:
            raise AttributeError("ADMIN_THEME value error: '%s'" % val)
        args = [get_admin_theme(v) for v in val]
        return theme.Theme(*args)
    elif isinstance(val, str):
        val = val.upper()
        if val not in theme.THEME_STYLES:
            raise AttributeError("ADMIN_THEME value error: '%s'" % val)
        val = theme.THEME_STYLES[val]
        return val if style else theme.Theme(val)
    elif isinstance(val, dict):
        val = theme.ThemeStyle(**val)
        return val if style else theme.Theme(val)
    else:
        raise AttributeError("ADMIN_THEME value error: '%r'" % val)


def perform_import(val, setting_name):
    """
    If the given setting is a string import notation,
    then perform the necessary import or imports.
    """
    if val is None:
        return None
    elif isinstance(val, str):
        return import_from_string(val, setting_name)
    elif isinstance(val, (list, tuple)):
        return [import_from_string(item, setting_name) for item in val]
    return val


def import_from_string(val, setting_name):
    """
    Attempt to import a class from a string representation.
    """
    try:
        return import_string(val)
    except ImportError as e:
        msg = "Could not import '%s' for API setting '%s'. %s: %s." % (val, setting_name, e.__class__.__name__, e)
        raise ImportError(msg)


class CoolSettings:
    """
    A settings object that allows Django-Cool settings to be accessed as
    properties. For example:
        from cool.settings import cool_settings
        print(cool_settings.FILTER_USE_SELECT)
    Any setting with string import paths will be automatically resolved
    and return the class, rather than the string literal.
    Note:
    This is an internal class that is only compatible with settings namespaced
    under the REST_FRAMEWORK name. It is not intended to be used by 3rd-party
    apps, and test helpers like `override_settings` may not work as expected.
    """
    def __init__(self, user_settings=None, defaults=None, import_strings=None):
        if user_settings:
            self._user_settings = user_settings
        self.defaults = defaults or DEFAULTS
        self.import_strings = import_strings or IMPORT_STRINGS
        self.admin_theme_string = "ADMIN_THEME"
        self._cached_attrs = set()

    @property
    def user_settings(self):
        if not hasattr(self, '_user_settings'):
            self._user_settings = getattr(settings, 'DJANGO_COOL', {})
        return self._user_settings

    def __getattr__(self, attr):
        if attr not in self.defaults:
            raise AttributeError("Invalid API setting: '%s'" % attr)

        try:
            # Check if present in user settings
            val = self.user_settings[attr]
        except KeyError:
            # Fall back to defaults
            val = self.defaults[attr]

        # Coerce import strings into classes
        if attr in self.import_strings:
            val = perform_import(val, attr)
        if attr == self.admin_theme_string:
            val = get_admin_theme(val)
        # Cache the result
        self._cached_attrs.add(attr)
        setattr(self, attr, val)
        return val

    def reload(self):
        for attr in self._cached_attrs:
            delattr(self, attr)
        self._cached_attrs.clear()
        if hasattr(self, '_user_settings'):
            delattr(self, '_user_settings')


cool_settings = CoolSettings(None, DEFAULTS, IMPORT_STRINGS)


def reload_cool_settings(*args, **kwargs):
    setting = kwargs['setting']
    if setting == 'DJANGO_COOL':
        cool_settings.reload()


setting_changed.connect(reload_cool_settings)
